flink-clients/src/main/java/org/apache/flink/client/deployment/AbstractJobClusterExecutor.java

```diff
@@ -66,7 +68,41 @@ public AbstractJobClusterExecutor(@Nonnull final ClientFactory clusterClientFact
 			LOG.info("Job has been submitted with JobID " + jobGraph.getJobID());
 
 			final boolean withShutdownHook = !configAccessor.getDetachedMode() && configAccessor.isShutdownOnAttachedExit();
-			return CompletableFuture.completedFuture(new ClusterClientJobClientAdapter<>(clusterClient, jobGraph.getJobID(), withShutdownHook));
+
+			if (withShutdownHook) {
+				Thread shutdownHook = ShutdownHookUtil.addShutdownHook(
+					clusterClient::shutDownCluster, clusterClient.getClass().getSimpleName(), LOG);
+
+				return CompletableFuture.completedFuture(new ClusterClientJobClientAdapter<ClusterID>(clusterClient, jobGraph.getJobID()) {
+					@Override
+					public void doClose() {
+						Throwable throwable = null;
+
+						try {
```

Comment:
Why not merging the two blocks like: 

```
try {
	ShutdownHookUtil.removeShutdownHook(shutdownHook, clusterClient.getClass().getSimpleName(), LOG);
     clusterClient.close();
} catch (Throwable t) {
	throwable = ExceptionUtils.firstOrSuppressed(t, throwable)
}

if (throwable != null) {
	ExceptionUtils.rethrow(throwable);
}
```

Also now the `clusterClient.close()` does not throw an exception so I am not sure if it is even needed to be in the `try-catch` block.

Reply:
Also in the rest of the method signatures, feel free to remove the `@Nonnull` annotations.

Reply:
After a close look I think it makes sense we close client as

```java
ShutdownHookUtil.removeShutdownHook(shutdownHook, clusterClient.getClass().getSimpleName(), LOG);
clusterClient.close();
```

for unchecked exception is unhandled.